File: 05_concurrency.ex
# elixir

defmodule Parallel do
  def pmap(collection, func) do
    collection
    |> Enum.map(&(Task.async(fn -> func.(&1) end)))
    |> Enum.map(&Task.await/1)
  end
end

defmodule Fibonacci do
  def calc(1), do: 1
  def calc(2), do: 1
  def calc(3), do: 2
  def calc(n) do
    calc(n-1) + calc(n-2)
  end
end


defmodule ConcurrencyExample do
  def concurrently_do_work(list) do
    Parallel.pmap(list, fn item -> do_some_work(item) end)
  end

  defp do_some_work(item) do
    Fibonacci.calc(item)
    IO.puts(item)
  end
end

start_time = :os.system_time(:millisecond)
ConcurrencyExample.concurrently_do_work([40,40,40,40])
end_time = :os.system_time(:millisecond)
diff = end_time - start_time
IO.puts("Took #{diff}") # => Took 1662

start_time = :os.system_time(:millisecond)
ConcurrencyExample.concurrently_do_work([40]) # => Takes the same amount of time as 4 iterations
end_time = :os.system_time(:millisecond)
diff = end_time - start_time
IO.puts("Took #{diff}") # => Took 1582

File: 05_concurrency.rb
# ruby

def fibonacci(n)
  if [0,1,2].include?(n)
    return 1
  end

  if n == 3
    return 2
  end

  return fibonacci(n-1) + fibonacci(n-2)
end

class ConcurrencyExample
  def concurrently_do_work(collection)
    threads = collection.map do |item|
      Thread.new do
        do_some_work(item)
      end
    end

    threads.map(&:join)
  end

  def do_some_work(item)
    x = fibonacci(item)
    puts "Fibonacci of #{item}: #{x}"
  end
end

x = ConcurrencyExample.new
start_time = Time.now
x.concurrently_do_work([35,35,35,35]) # => takes 4x as long as one iteration
end_time = Time.now

diff = end_time - start_time
puts "Took #{diff}" # => Took 5.125619

start_time = Time.now
x.concurrently_do_work([35])
end_time = Time.now
diff = end_time - start_time
puts "Took #{diff}" # => Took 1.252302
